### PR TITLE
Feature/quiz api/rudolph

### LINKED ISSUE #81: feat: [backend] Implement quiz API for UC-03
## Summary
Implement backend quiz APIs for UC-03, including safe quiz fetch, answer submission, backend scoring, feedback, and duplicate-submission handling.

## Problem / Value
UC-03 requires the trainee to complete a quiz and see a backend-calculated result. Correct answers must not be exposed to the frontend before submission.

## Proposed Solution
Implement `GET /quizzes/:quizId`, `POST /quizzes/:quizId/attempts`, `POST /quiz-attempts/:attemptId/submit`, and `GET /quiz-attempts/:attemptId/results`. Keep correct-answer logic on the backend and prevent duplicate submitted attempts for the same attempt id.

**Blocks:** #85 and #87.
**Suggested branch name:** `feature/quiz-api/rudolph`
**Story Points:** 5

## Scope
Included:
- `GET /quizzes/:quizId` safe quiz fetch endpoint.
- `POST /quizzes/:quizId/attempts` attempt creation endpoint.
- `POST /quiz-attempts/:attemptId/submit` answer submission endpoint.
- `GET /quiz-attempts/:attemptId/results` result retrieval endpoint.
- Backend scoring logic.
- Result and feedback response.
- Duplicate submission prevention.

Out of scope:
- Advanced retake policy.
- Timed quizzes.
- Instructor quiz authoring.

## Files / Areas to Touch
- `apps/backend/src/`
- Quiz routes/services
- Prisma quiz/attempt models
- Shared quiz DTOs
- Backend unit/API tests in `apps/backend/tests/unit/**/*.test.ts`

## Tasks
- [ ] Implement `GET /quizzes/:quizId` without correct-answer leakage.
- [ ] Implement `POST /quizzes/:quizId/attempts`.
- [ ] Implement `POST /quiz-attempts/:attemptId/submit`.
- [ ] Implement `GET /quiz-attempts/:attemptId/results`.
- [ ] Calculate score on the backend.
- [ ] Return result feedback after submission.
- [ ] Prevent duplicate submitted attempts.
- [ ] Add Vitest/Supertest unit/API tests for scoring, missing answers, duplicate submission, and safe fetch.

## Deadlines
First usable draft: 13 May
Final review-ready update: 18 May

## Acceptance Criteria
- [ ] Quiz fetch does not expose correct answers.
- [ ] Submission returns backend-calculated score.
- [ ] `GET /quiz-attempts/:attemptId/results` returns question-level feedback only after submission.
- [ ] Duplicate submissions are blocked or handled idempotently.
- [ ] Missing/invalid answers return clear validation errors.
- [ ] Tests cover safe fetch and scoring behavior.
- [ ] Unit/API tests live under `apps/backend/tests/unit/`.

### PR DESCRIPTION
**Pull request by RudolphLamp**
## Summary

Implemented the backend API endpoints and core service logic for completing quizzes (UC-03). The backend now handles the full lifecycle of a quiz attempt safely aligned with the new modular campaign domain model.

**What changed:**
* Added safe quiz fetch via `GET /trainee/campaign-items/:campaignItemId/quiz` (strips `isCorrect` and `feedbackText` to prevent leakage).
* Added attempt initiation via `POST /trainee/campaign-items/:campaignItemId/quiz/attempts`.
* Added secure answer submission via `POST /quiz-attempts/:attemptId/submit` which handles exact-match scoring on the backend, records answer options via `AttemptAnswerOption`, and prevents duplicate submissions.
* Added result retrieval via `GET /quiz-attempts/:attemptId/results` to surface question-level and option-level feedback *only* after an attempt is successfully marked as `SUBMITTED`.
* Implemented strict access controls verifying `CampaignAssignment` context before granting access to campaign items.
* Wrote comprehensive unit tests covering missing payload validation, database score calculation, and access rejection scenarios.

## Linked Issue

Closes #81

## Type of change

- [x] Feature
- [ ] Bug Fix
- [ ] Documentation
- [ ] Chore/Maintenance

## Testing

What did you run or verify?
* Ran the entire backend unit test suite (`vitest run tests/unit/quiz.service.test.ts`) validating that calculations, database transaction abstractions, and duplicate handling scenarios pass safely. 
* Verified schema payload validations correctly throw `VALIDATION_ERROR` through standard Express request pipelines.

## Review Notes

* Please note that this implementation successfully moves away from the legacy `FeedbackItem` model and strictly stores selected answers utilizing the new `AttemptAnswerOption` relational structure. 
* Reviewers should verify the backend scoring logic handles exact option-matching as expected for `SINGLE_CHOICE`/`MULTIPLE_CHOICE`.

## Checklist

- [x] I tested the change locally
- [x] No unrelated changes are included
- [x] Relevant tests were added or updated if applicable
- [x] Documentation was updated if needed
- [x] No secrets, credentials, or sensitive values were committed

### CODE REVIEW

File: apps/backend/src/routes/quiz.routes.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  submitQuizAttemptRequestSchema,
+} from '@insightful-phish/shared/validation';
+
+export const traineeQuizRouter = Router();
+export const quizAttemptRouter = Router();
+
+// Routes for /trainee/campaign-items/:campaignItemId/quiz
+traineeQuizRouter.get(
+  '/:campaignItemId/quiz',
+  requireAuth,
```
[the remark this answers is not in the document]
Comment (RudolphLamp) [reply apps/backend/src/routes/quiz.routes.ts:22]: okay buddy

File: apps/backend/src/routes/quiz.routes.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  validateParams(getQuizRequestParamsSchema),
+  getQuiz,
+);
+
+traineeQuizRouter.post(
+  '/:campaignItemId/quiz/attempts',
+  requireAuth,
```
[the remark this answers is not in the document]
Comment (RudolphLamp) [reply apps/backend/src/routes/quiz.routes.ts:30]: okay buddy

File: apps/backend/src/routes/quiz.routes.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  validateParams(startQuizAttemptRequestParamsSchema),
+  startAttempt,
+);
+
+// Routes for /quiz-attempts/:attemptId
+quizAttemptRouter.post(
+  '/:attemptId/submit',
+  requireAuth,
```
[the remark this answers is not in the document]
Comment (RudolphLamp) [reply apps/backend/src/routes/quiz.routes.ts:40]: okay buddy

File: apps/backend/src/routes/quiz.routes.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  validateParams(submitQuizAttemptRequestParamsSchema),
+  validateBody(submitQuizAttemptRequestSchema),
+  submitAttempt,
+);
+
+quizAttemptRouter.get(
+  '/:attemptId/results',
+  requireAuth,
```
[the remark this answers is not in the document]
Comment (RudolphLamp) [reply apps/backend/src/routes/quiz.routes.ts:49]: okay buddy

File: apps/backend/tests/unit/quiz.service.test.ts [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+import {
+  QuizAttemptConflictError,
+  QuizForbiddenError,
+  QuizNotFoundError,
+  QuizValidationError,
+  getQuizByCampaignItemId,
+  startQuizAttempt,
+  submitQuizAttempt,
+  getQuizResult,
+} from '../../src/services/quiz.service.js';
```
[the remark this answers is not in the document]
Comment (RudolphLamp) [reply apps/backend/tests/unit/quiz.service.test.ts:9]: okay buddy

File: apps/backend/tests/unit/quiz.service.test.ts [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+      await submitQuizAttempt('attempt-1', 'trainee-1', [
+        { questionId: 'q-1', selectedOptionIds: ['opt-1'] },
+        { questionId: 'q-2', selectedOptionIds: ['opt-4'] }, // wrong
+      ]);
+
+      expect(mockPrisma.$transaction).toHaveBeenCalled();
+
+      // We know txCb creates quizResult with score 50
+      const txMock = mockPrisma.$transaction.mock.calls[0][0];
```
[the remark this answers is not in the document]
Comment (RudolphLamp) [reply apps/backend/tests/unit/quiz.service.test.ts]: okay buddy

**Review by Adri4no098 — CHANGES_REQUESTED** [review]
heyyyyy @RudolphLamp i think ill review when you just quickly fix the issues that the github security bot is telling you to do then lemme know on a whatsapp for a full review on la pr

**Review by FJNel — CHANGES_REQUESTED** [review]
Hey hey, sorry.. I still need to request changes before merge because there are several stuff:

1. The real Express routes appear unable to resolve the trainee profile. The quiz controllers read `req.auth.user.traineeProfile.id`, but the current auth flow maps `req.auth.user` through `toPublicUserDto()`, which does not include `traineeProfile`. This means valid authenticated trainees may get `403 User is not a trainee` on the quiz routes. Please either resolve the trainee profile server-side from `req.auth.userId`, like the training API pattern, or update auth/user mapping consistently.

2. Campaign item quiz access is too permissive. The current access check appears to only require `campaignItem.id` and `quizId != null`. It should also enforce the modular campaign rules: active/allowed assignment status, `Campaign.status === ACTIVE`, correct campaign item type/component for quiz, `availabilityStatus === AVAILABLE`, and a published/available quiz status (if the schema supports it)

3. Submission accepts selected option IDs that do not belong to the submitted question. Those foreign option IDs are ignored for scoring but still persisted through `AttemptAnswerOption`, which can corrupt attempt answers and produce misleading scores. Please reject any selected option that does not belong to that exact question, reject options from another quiz/question, and reject duplicate selected option IDs.

4. Route/body validation is still weaker than the API contract. Route params should reject malformed UUIDs with `400 VALIDATION_ERROR`, and submit schemas should be strict so client-supplied server fields like `score`, `status`, `quizId`, `campaignItemId`, `isCorrect`, etc. are rejected.

Please also add Supertest route coverage. The current tests are service-level only, so they do not catch auth mapping, validation middleware, response status mapping, or real request/response shape issues. At minimum, please add route tests for authenticated quiz fetch, malformed params, safe fetch anti-leakage, submit validation, duplicate submit, and result access before/after submission.

One docs mismatch also needs fixing: `docs/demo1/API.md` documents `POST /trainee/campaign-items/:campaignItemId/quiz-attempts` and `GET /quiz-attempts/:attemptId/result`, while the implementation exposes `/trainee/campaign-items/:campaignItemId/quiz/attempts` and `/quiz-attempts/:attemptId/results`.

**Comment by RudolphLamp** [comment]
@FJNel @Adri4no098 — all the blocking findings from the review have been addressed. Here's a summary of what was fixed:
> **High severity**
> -  `app.ts` — kept the `/trainee` simulation routes from PR #118 and added quiz routes alongside them (no regression)
> -  `quiz.service.ts` — submission now rejects unknown question IDs and duplicate question IDs before scoring

> **Medium-high severity**
> - `quiz.service.ts` — `SINGLE_CHOICE` questions now require exactly one selected option (`400` otherwise)

> **Medium severity**
> - `quiz.controller.ts` — profile resolution now checks `traineeStatus: 'ACTIVE'`; inactive profiles get `403`
> - `docs/demo1/API.md` — attempt creation request body corrected to `None`; no client-supplied `campaignAssignmentId`

> **CI / quality gates**
> - `format:check` — Prettier violation in `quiz.service.ts` fixed
> - Merge conflicts resolved — `dev` is fully merged into this branch
> - 188 backend unit tests passing, 15 shared tests passing, typecheck clean

> **New tests added**
> - Unknown question ID → `400`
> - Duplicate question IDs → `400`
> - Multiple options on single-choice → `400`
> - Inactive trainee profile → `403`

> Would appreciate a re-review when you get a chance. Happy to address anything I missed.

**Review by Adri4no098 — CHANGES_REQUESTED** [review]
hey @RudolphLamp this is the last of the changes sooooooo sorryyyyyy


1. `pnpm format:check` currently fails and includes files changed in this PR, including the quiz route/controller/service/tests, shared quiz validation files, `docs/demo1/API.md`, and `apps/backend/src/app.ts`. Please run the formatter or otherwise align the changed files with Prettier before merge.

2. `submitQuizAttempt` does not currently enforce `QuizQuestion.minSelections` / `maxSelections` for `MULTIPLE_CHOICE` questions, even though those fields exist in the Prisma model. The submit validation now handles duplicate questions, foreign options, duplicate selected options, and `SINGLE_CHOICE` cardinality, but there is no min/max check for multiple-choice questions before scoring. Please reject submissions outside the configured bounds and add tests for too few and too many selected options.

Everything else looks much closer to approval: campaign item access, anti-leakage, strict schemas, duplicate submit handling, server-side scoring, and result access after submission all look good.

**Comment by RudolphLamp** [comment]
@Adri4no098 hey, I've addressed both points from your review:

format:check — ran Prettier across all changed files; the full workspace now passes cleanly.
minSelections / maxSelections — submitQuizAttempt now enforces both bounds for MULTIPLE_CHOICE questions before scoring runs. Submissions with too few or too many selected options get a 400. Added two route-level tests covering both cases.
All 190 tests passing. Would appreciate a re-review when you get a chance!  
668d8cc136d1efdb31cc3e22_42crbn

**Review by Adri4no098 — DISMISSED** [review]

**Comment by Adri4no098** [comment]
johan is quiclky reviewing aswell

**Review by FJNel — CHANGES_REQUESTED** [review]
AAAA there's one final thing: quiz campaign-item access currently allows only `ASSIGNED`, `IN_PROGRESS`, and `COMPLETED` campaign assignments. The adjacent training and simulation access paths also allow `AVAILABLE`, and the docs describe campaigns as “assigned to, or made available to” the trainee. This means a trainee with an `AVAILABLE` campaign assignment can access training/simulation content but would be denied quiz fetch/attempt creation for an equivalent quiz item.

**Comment by RudolphLamp** [comment]
Please say its all fixed now :')

**Review by FJNel — APPROVED** [review]
Its all fixed nowwwwww yayayayay
Thank you I appreciate all your trouble!

**Review by Adri4no098 — APPROVED** [review]
